src/grpc.js

```diff
@@ -0,0 +1,259 @@
+import EventEmitter from 'events'
+import StateMachine from 'javascript-state-machine'
+import createDebug from 'debug'
+import { status } from '@grpc/grpc-js'
+import { validateHost } from './utils'
+import { WalletUnlocker, Lightning, Autopilot, ChainNotifier, Invoices, Router, Signer, WalletKit } from './services'
+import registry from './registry'
+
+const debug = createDebug('lnrpc:grpc')
+
+const WALLET_STATE_LOCKED = 'WALLET_STATE_LOCKED'
+const WALLET_STATE_ACTIVE = 'WALLET_STATE_ACTIVE'
+
+/**
+ * Lnd gRPC service wrapper.
+ * @extends EventEmitter
+ */
+class LndGrpc extends EventEmitter {
+  constructor(options = {}) {
+    super()
+    debug(`Initializing LndGrpc with config: %o`, options)
+    this.options = options
+
+    // Define state machine.
+    this.fsm = new StateMachine({
+      init: 'ready',
+      transitions: [
+        { name: 'activateWalletUnlocker', from: ['ready', 'active'], to: 'locked' },
+        { name: 'activateLightning', from: ['ready', 'locked'], to: 'active' },
+        { name: 'disconnect', from: ['locked', 'active'], to: 'ready' },
+      ],
+      methods: {
+        onBeforeActivateWalletUnlocker: this.onBeforeActivateWalletUnlocker.bind(this),
+        onBeforeActivateLightning: this.onBeforeActivateLightning.bind(this),
+        onBeforeDisconnect: this.onBeforeDisconnect.bind(this),
+        onAfterDisconnect: this.onAfterDisconnect.bind(this),
+      },
+    })
+
+    // Define services.
+    this.supportedServices = [WalletUnlocker, Lightning, Autopilot, ChainNotifier, Invoices, Router, Signer, WalletKit]
+    this.services = {}
+
+    // Instantiate services.
+    this.supportedServices.forEach(Service => {
+      const instance = new Service(this.options)
+      this.services[instance.serviceName] = instance
+    })
+  }
+
+  // ------------------------------------
+  // FSM Proxies
+  // ------------------------------------
+
+  is(...args) {
+    return this.fsm.is(args)
+  }
+  can(...args) {
+    return this.fsm.can(args)
+  }
+  observe(...args) {
+    return this.fsm.observe(args)
+  }
+  get state() {
+    return this.fsm.state
+  }
+
+  async connect() {
+    debug(`Connecting to lnd gRPC service`)
+
+    // Verify that the host is valid.
+    const { host } = this.options
+    await validateHost(host)
+
+    // Probe the services to determine the wallet state.
+    const walletState = await this.determineWalletState()
```

Comment:
`determineWalletState` throws in some cases. Where does that situation is supposed to be handled?

Reply:
Any error thrown by `determineWalletState` should bubble up to the caller. So in this case, `connect` should throw the error back to the end user.